src/lib.rs

```diff
@@ -1311,6 +1315,112 @@ where
     Some(Ratio::new(n1, d1))
 }
 
+#[cfg(all(feature = "bigint", feature = "std"))]
+impl<T: Clone + Integer + Signed + ToPrimitive + ToBigInt> Ratio<T> {
+    /// Converts the ratio to an `f64`.
+    ///
+    /// The value is guaranteed to be the closest `f64` approximating the ratio. This always
+    /// returns a value, unless the conversion to BigInt for the numerator or denominator fails.
+    ///
+    /// ```
+    /// use num_rational::Rational64;
+    /// assert_eq!(0.5f64, Rational64::new(1, 2).to_f64().unwrap());
+    /// ```
+    pub fn to_f64(&self) -> Option<f64> {
+        assert_eq!(
+            std::f64::RADIX,
+            2,
+            "only floating point implementations with radix 2 are supported"
+        );
+
+        // Upper bound to the range of exactly-representable ints in an f64.
+        const MAX_EXACT_INT: u64 = 1u64 << std::f64::MANTISSA_DIGITS;
+
+        let numer: T = self.numer().abs();
+        let denom: T = self.denom().abs();
+        let sign: T = self.numer().signum() * self.denom().signum();
+        let flo_sign = if sign.is_negative() { -1.0 } else { 1.0 };
+
+        if numer.is_zero() {
+            return Some(0.0 * flo_sign);
+        }
+
+        // Fast track: both sides can losslessly be converted to f64s. In this case, letting the
+        // FPU do the job is faster and easier. In any other case, converting to f64s may lead
+        // to an inexact result: https://stackoverflow.com/questions/56641441/.
+        if let (Some(n), Some(d)) = (numer.to_u64(), denom.to_u64()) {
+            if n < MAX_EXACT_INT && d < MAX_EXACT_INT {
+                return Some(numer.to_f64().unwrap() / denom.to_f64().unwrap() * flo_sign);
+            }
+        }
+
+        // Otherwise, the goal is to obtain a quotient with at least 55 bits. 53 of these bits will
+        // be used as the mantissa of the resulting float, and the remaining two are for rounding.
+        // There's an error of up to 1 on the number of resulting bits, so we may get either 55 or
+        // 56 bits.
+        let mut numer = otry!(numer.to_bigint());
+        let mut denom = otry!(denom.to_bigint());
+        let (is_diff_positive, absolute_diff) = match numer.bits().checked_sub(denom.bits()) {
+            Some(diff) => (true, diff),
+            None => (false, denom.bits() - numer.bits()),
+        };
+
+        // Filter out overflows and underflows. After this step, the signed difference fits in an
+        // isize.
+        if is_diff_positive && absolute_diff > std::f64::MAX_EXP as usize {
+            return Some(std::f64::INFINITY * flo_sign);
+        }
+        if !is_diff_positive
+            && absolute_diff > -std::f64::MIN_EXP as usize + std::f64::MANTISSA_DIGITS as usize + 1
+        {
+            return Some(0.0 * flo_sign);
+        }
+        let diff = if is_diff_positive {
+            absolute_diff as isize
```

Comment:
maybe use `isize::try_from(absolute_diff)` here (and below). I can't think of a situation where what you're doing would fail, but it feels like better defensive programming.

Reply:
Looks like `try_from` was only stabilized recently, so I used rust-num's `to_isize()` instead.